wolfhsm/wh_message_custom.h

```diff
@@ -0,0 +1,79 @@
+#ifndef WH_MESSAGE_CUSTOM_H_
+#define WH_MESSAGE_CUSTOM_H_
+
+#include <stdint.h>
+
+#define WH_MESSAGE_CUSTOM_BUF_SIZE (256)
+
+/* Type indicator for custom request/response messages. Indicates how
+ * to interpret whMessageCustomData */
+typedef enum {
+    /* message types reserved for internal usage*/
+    WH_MESSAGE_CUSTOM_TYPE_QUERY      = 0,
+    WH_MESSAGE_CUSTOM_TYPE_DMA32      = 1,
+    WH_MESSAGE_CUSTOM_TYPE_DMA64      = 2,
+    WH_MESSAGE_CUSTOM_TYPE_RESERVED_3 = 3,
+    WH_MESSAGE_CUSTOM_TYPE_RESERVED_4 = 4,
+    WH_MESSAGE_CUSTOM_TYPE_RESERVED_5 = 5,
+    WH_MESSAGE_CUSTOM_TYPE_RESERVED_6 = 6,
+    WH_MESSAGE_CUSTOM_TYPE_RESERVED_7 = 7,
+    /* User-defined types start from here, up to UINT32_MAX */
+    WH_MESSAGE_CUSTOM_TYPE_USER_DEFINED_START = 8,
+} whMessageCustom_Type;
+
+
+/* union providing some helpful abstractions for passing pointers in/out of
+ * custom callbacks on top of a raw data buffer */
+typedef union {
+    /* pointer/size pairs for 32-bit systems */
+    struct {
+        uint32_t client_addr;
+        uint32_t client_sz;
+        uint32_t server_addr;
+        uint32_t server_sz;
+    } dma32;
+    /* pointer/size pairs for 64-bit systems */
+    struct {
+        uint64_t client_addr;
+        uint64_t client_sz;
+        uint64_t server_addr;
+        uint64_t server_sz;
+    } dma64;
+    /* raw data buffer for user-defined schema */
+    struct {
+        uint8_t data[WH_MESSAGE_CUSTOM_BUF_SIZE];
+    } buffer;
+} whMessageCustom_Data;
+
+/* request message to the custom server callback */
+typedef struct {
+    uint16_t             id;   /* indentifier of registered callback  */
```

Comment:
@billphipps should I just make it uint32_t instead? It is easier to grok with that field at the top

Reply:
Since we are trying to avoid a marshaling scheme and I still think the packed attribute is evil, we need to be careful to set up the message structs so any compiler will put it the same way.  Your choice if you want to make it a uint32_t or if reordering is easier.   Just make sure you always do a sizeof(struct) check on receipt of the request and response to ensure neither side had any packing shenanigans.